[MD-929] StakerRewards are now upgradeable and use storage.

steven2308 opened the pull request.

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -155,9 +139,13 @@ contract ODefaultStakerRewards is
-        NETWORK = params.network;
+
+        StakerRewardsStorage storage $ = _getStakerRewardsStorage();
+        $.VAULT = params.vault;
```
I'm thinking that probably these should be immutable or what you think? I mean once a staker reward is set the network and vault won't change. Actually if they would change it would be problematic innit?

NindoK replied:
Especially since now dont have an implementation but we just deploy the contract.

steven2308 replied:
Agree. Moved them.
I'd like to make all (EVERYWHERE) immutable uppercase, any objections? I think with that we can even remove the `i_` prefix.  I don't think we are storing any constant in storage (would it even make sense?) so we can easily tell them apart.

steven2308 replied:
I'll do this in another PR.

NindoK replied:
I'd keep the i_ for being consistent across all other contracts

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -272,15 +270,17 @@ contract ODefaultStakerRewards is
-        if (_s_activeSharesCache[epoch] == 0) {
-            uint256 activeShares_ = IVault(VAULT).activeSharesAt(epochTs, activeSharesHint);
-            uint256 activeStake_ = IVault(VAULT).activeStakeAt(epochTs, activeStakeHint);
+        StakerRewardsStorage storage $ = _getStakerRewardsStorage();
```
I feel we are pasing down the storage too many times. Wouldn't be better to just pass it down as input to the function?

NindoK replied:
In all functions

steven2308 replied:
Passed for the 3 methods were it made sense. All others use it just in the same method.

steven2308 replied:
Nevermind, 2 of them produced stack too deep. I can try further tomorrow.

NindoK replied:
Ye it's just that it feels like a waste of gas. If it's possible fine otherwise whatever

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -314,54 +316,70 @@ contract ODefaultStakerRewards is
 
         // If there are no rewards to claim, revert
         if (rewardsToClaim == 0) {
             revert ODefaultStakerRewards__NoRewardsToClaim();
```
Put this directly in _getRewardsToClaim

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -314,54 +316,70 @@ contract ODefaultStakerRewards is
-            _claimRewardsPerEpoch(rewardsToClaim, rewardsPerEpoch, lastUnclaimedReward_, epoch, activeSharesOfHints);
-
-        s_lastUnclaimedReward[msg.sender][epoch][tokenAddress] = lastUnclaimedReward_;
+        $.lastUnclaimedReward[msg.sender][epoch][tokenAddress] = lastUnclaimedReward_ + rewardsToClaim;
```
This update can be made dirctly inside `_getRewardsToClaimAndAmount`

NindoK replied:
So I would do: 

`lastUnclaimedReward_= lastUnclaimedReward_  + rewardsToClaim;`
`$.lastUnclaimedReward[msg.sender][epoch][tokenAddress] = lastUnclaimedReward_;`

NindoK replied:
This is linked also to the comment below

steven2308 replied:
All done.

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -314,54 +316,70 @@ contract ODefaultStakerRewards is
 
         emit ClaimRewards(
-            NETWORK, tokenAddress, msg.sender, epoch, recipient, lastUnclaimedReward_, rewardsToClaim, amount
+            $.NETWORK, tokenAddress, msg.sender, epoch, recipient, lastUnclaimedReward_, rewardsToClaim, amount
```
lastUnclaimedReward_ should be lastUnclaimedReward_ + rewardsToClaim in this case

steven2308 replied:
I had this doubt, on the purpose of the event. But this behavior you mention makes more sense.

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -407,13 +427,74 @@ contract ODefaultStakerRewards is
+
+    function _authorizeUpgrade(
+        address newImplementation
+    ) internal override onlyRole(DEFAULT_ADMIN_ROLE) {}
```
👍

NindoK commented on test/unit/Rewards.t.sol:
```diff
@@ -73,6 +73,9 @@ contract RewardsTest is Test {
     uint48 public constant OPERATOR_SHARE = 2000;
 
+    // keccak256(abi.encode(uint256(keccak256("tanssi.rewards.ODefaultStakerRewards.v1")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 constant MAIN_STORAGE_LOCATION = 0xe07cde22a6017f26eee680b6867ce6727151fb6097c75742cbe379265c377400;
```
You can directly access this from a contract, so we dont need to maintain this

steven2308 commented:
Addressed all the comments except passing down storage where possible. It only was possible in 3 places but 2 of them resulted in stack too deep. I can keep trying tomorrow but wouldn't block the merge and deploy for that.

Also agreed in making vault and network immutable instead. I do still include them in the init params since it's more convenient, but let me know your thoughts @NindoK

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -252,18 +274,24 @@ contract ODefaultStakerRewards is
+        emit DistributeRewards(i_network, tokenAddress, eraIndex, epoch, amount, data);
+    }
+
+    function _transferAndCheckAmount(address tokenAddress, uint256 amount) private returns (uint256) {
```
Just add here the named returns?

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -272,15 +300,17 @@ contract ODefaultStakerRewards is
         bytes memory activeSharesHint,
         bytes memory activeStakeHint
```
Here shouldn't cause stack too deep right?

steven2308 replied:
It causes the stack too deep on the caller. I tried again and discovered I can do one of them but not 2. :man_shrugging: 
This and the next one are called from the same method.

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -290,10 +320,12 @@ contract ODefaultStakerRewards is
         // And distribute the rest to the stakers
```
Same here

steven2308 replied:
It causes the stack too deep on the caller.

NindoK replied:
Whaaaaattt 😨

NindoK commented on src/contracts/rewarder/ODefaultStakerRewards.sol:
```diff
@@ -334,34 +382,34 @@ contract ODefaultStakerRewards is
-            IERC20(tokenAddress).safeTransfer(recipient, amount);
+        // If there are no rewards to claim, revert
+        if (rewardsToClaim == 0) {
+            revert ODefaultStakerRewards__NoRewardsToClaim();
```
Should go after 376
